File: impl-go/jcs.go
package main

import (
	"fmt"
	"sort"
	"strconv"
	"strings"
	"unicode/utf16"
	"unicode/utf8"
)

// jcs implements RFC 8785 (JSON Canonicalization Scheme), byte-for-byte identical to impl-py's
// jcs() and src/jcs.ts:
//   - integers serialized as their decimal form (floats are impossible here — the parser rejects them),
//   - object keys sorted by UTF-16 code units (NOT UTF-8 byte order),
//   - RFC-8785 string escaping (control chars escaped, every other code point emitted literally),
//   - no Unicode normalization.
// The canonical string is the exact input to sha256 for the receipt hash + signing preimage.
func jcs(v *Value) (string, error) {
	switch v.Kind {
	case KindNull:
		return "null", nil
	case KindBool:
		if v.Bool {
			return "true", nil
		}
		return "false", nil
	case KindInt:
		return strconv.FormatInt(v.Int, 10), nil
	case KindString:
		return jcsString(v.Str)
	case KindArray:
		var b strings.Builder
		b.WriteByte('[')
		for i, e := range v.Arr {
			if i > 0 {
				b.WriteByte(',')
			}
			s, err := jcs(e)
			if err != nil {
				return "", err
			}
			b.WriteString(s)
		}
		b.WriteByte(']')
		return b.String(), nil
	case KindObject:
		keys := make([]string, 0, len(v.Obj))
		for k := range v.Obj {
			keys = append(keys, k)
		}
		sortUTF16(keys)
		var b strings.Builder
		b.WriteByte('{')
		for i, k := range keys {
			if i > 0 {
				b.WriteByte(',')
			}
			ks, err := jcsString(k)
			if err != nil {
				return "", err
			}
			b.WriteString(ks)
			b.WriteByte(':')
			vs, err := jcs(v.Obj[k])
			if err != nil {
				return "", err
			}
			b.WriteString(vs)
		}
		b.WriteByte('}')
		return b.String(), nil
	}
	return "", fmt.Errorf("unsupported value type")
}

// jcsString mirrors impl-py jcs_string / src/jcs.ts serializeString.
func jcsString(s string) (string, error) {
	// Parser already guarantees valid UTF-8 with no lone surrogates; this guard mirrors impl-py's
	// s.encode("utf-8") / isWellFormed() check for defense in depth.
	if !utf8.ValidString(s) {
		return "", fmt.Errorf("unpaired surrogate / invalid UTF-8 in string")
	}
	var b strings.Builder
	b.WriteByte('"')
	for _, ch := range s {
		switch ch {
		case '"':
			b.WriteString("\\\"")
		case '\\':
			b.WriteString("\\\\")
		case '\b':
			b.WriteString("\\b")
		case '\f':
			b.WriteString("\\f")
		case '\n':
			b.WriteString("\\n")
		case '\r':
			b.WriteString("\\r")
		case '\t':
			b.WriteString("\\t")
		default:
			if ch < 0x20 {
				b.WriteString(fmt.Sprintf("\\u%04x", ch))
			} else {
				b.WriteRune(ch)
			}
		}
	}
	b.WriteByte('"')
	return b.String(), nil
}

// sortUTF16 sorts keys by UTF-16 code-unit order, equivalent to impl-py's
// sorted(keys, key=lambda k: k.encode("utf-16-be")) and JS's default string sort (RFC 8785).
func sortUTF16(keys []string) {
	sort.Slice(keys, func(a, b int) bool { return lessUTF16(keys[a], keys[b]) })
}

func lessUTF16(a, b string) bool {
	ua := utf16.Encode([]rune(a))
	ub := utf16.Encode([]rune(b))
	n := len(ua)
	if len(ub) < n {
		n = len(ub)
	}
	for i := 0; i < n; i++ {
		if ua[i] != ub[i] {
			return ua[i] < ub[i]
		}
	}
	return len(ua) < len(ub)
}

File: packages/approval-artifacts/src/jcs.ts
/**
 * RFC 8785 (JSON Canonicalization Scheme) — hardened, PORTED BYTE-FOR-BYTE from
 * `noa-receipt/src/jcs.ts` (identical algorithm; the same file `packages/signer-core/src/jcs.ts`
 * already mirrors, with the same rationale). This package ports rather than imports for the same
 * reason signer-core does — to stay a self-contained, install-order-free conformance package with
 * ZERO runtime dependencies (only `node:crypto` via ./crypto.ts). `test/parity.test.ts` proves this
 * copy is byte-identical to the reference for the shared receipt/side-artifact inputs.
 *
 * The canonical byte-form is the input to every hash (the signing preimage AND `refHash`), so any
 * producer/verifier disagreement on those bytes is a silent forgery channel. The implementation is
 * deliberately STRICT and SMALL: floats / non-finite / unsafe-integer numbers are REJECTED (side
 * artifacts, like receipts, use integers only); object keys sorted by UTF-16 code units (RFC 8785);
 * control chars escaped, all other code points emitted literally as UTF-8; undefined / functions /
 * symbols / bigint REJECTED. Inputs MUST already be NFC-normalized by the producer.
 */

export class JcsError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "JcsError";
  }
}

/** Canonicalize a JSON-compatible value to its RFC 8785 byte-string form. */
export function canonicalize(value: unknown): string {
  return serialize(value, 0);
}

/** Single source of truth for nesting depth. Mirrors noa-receipt/src/jcs.ts's MAX_DEPTH. */
export const MAX_DEPTH = 64;

function serialize(v: unknown, depth: number): string {
  if (depth > MAX_DEPTH) throw new JcsError("max nesting depth exceeded");

  if (v === null) return "null";

  const t = typeof v;

  if (t === "boolean") return v ? "true" : "false";

  if (t === "number") {
    const n = v as number;
    if (!Number.isFinite(n)) throw new JcsError("non-finite number not allowed");
    if (!Number.isInteger(n)) throw new JcsError("non-integer (float) not allowed in receipts");
    if (!Number.isSafeInteger(n)) throw new JcsError("integer outside safe range not allowed");
    if (Object.is(n, -0)) return "0";
    return n.toString();
  }

  if (t === "bigint") throw new JcsError("bigint not allowed");
  if (t === "string") return serializeString(v as string);

  if (Array.isArray(v)) {
    let out = "[";
    for (let i = 0; i < v.length; i++) {
      if (i > 0) out += ",";
      out += serialize(v[i], depth + 1);
    }
    return out + "]";
  }

  if (t === "object") {
    const obj = v as Record<string, unknown>;
    // Sort by UTF-16 code units (RFC 8785). JS default sort on strings does exactly this.
    const keys = Object.keys(obj).sort();
    let out = "{";
    let first = true;
    for (const k of keys) {
      const val = obj[k];
      if (val === undefined) throw new JcsError(`undefined value at key "${k}" not allowed`);
      if (!first) out += ",";
      first = false;
      out += serializeString(k) + ":" + serialize(val, depth + 1);
    }
    return out + "}";
  }

  throw new JcsError(`unsupported value type: ${t}`);
}

function serializeString(s: string): string {
  // Reject unpaired surrogates: the UTF-8 hashing step would silently map every lone surrogate to
  // U+FFFD, collapsing 2048 distinct code points to one hash bucket — a forgery channel.
  if (!s.isWellFormed()) throw new JcsError("unpaired surrogate in string not allowed");
  let out = '"';
  for (const ch of s) {
    switch (ch) {
      case '"':
        out += '\\"';
        break;
      case "\\":
        out += "\\\\";
        break;
      case "\b":
        out += "\\b";
        break;
      case "\f":
        out += "\\f";
        break;
      case "\n":
        out += "\\n";
        break;
      case "\r":
        out += "\\r";
        break;
      case "\t":
        out += "\\t";
        break;
      default: {
        const code = ch.codePointAt(0)!;
        if (code < 0x20) {
          out += "\\u" + code.toString(16).padStart(4, "0");
        } else {
          out += ch;
        }
      }
    }
  }
  return out + '"';
}
